src/Components/Components/src/RenderTree/RenderTreeBuilder.cs

```diff
@@ -590,13 +644,42 @@ public void Clear()
 
         private void Append(in RenderTreeFrame frame)
         {
+            var frameType = frame.FrameType;
+            if (frameType == RenderTreeFrameType.Attribute)
+            {
+                ClearAttributesWithName(frame.AttributeName);
+            }
+
             _entries.Append(frame);
 
-            var frameType = frame.FrameType;
             if (frameType != RenderTreeFrameType.Attribute)
             {
                 _lastNonAttributeFrameType = frame.FrameType;
             }
         }
+
+        private void ClearAttributesWithName(string name)
```

Comment:
Options I see for this:
1. What I did here (clear dupes of each attribute as it's added) - simple and O(N^2)
2. Clear dupes of *all* attributes for an element/component when it's closed - somewhat simple and also O(N^2)
3. Use a HashSet - simple, requires extra data/hashing and still kinda O(N^2) before removing from an array is O(N)

Basically I felt like keeping the data in the *always consistent* state was the best thing overall. Without doing dedicated perf work around this specific feature, I don't have a clear insight into which of these are the best. That seems like a good thing to do some other time.

Reply:
I'm not at all keen on the `ClearAttributesWithName` technique. It might be just about tolerable if it only applied when we know you're splatting, but to do it in all cases seems very troublesome. Adding a linear search to find matches, including the case-insensitive string comparisons, makes all parameters everywhere more expensive.

So most importantly, I think we only need to handle duplicates at all once we observe the first call to `AddMultipleAttributes` within a given container.

As for the deduplication implementation, what do you think of the following slight variation of combining your options 2 and 3?

 * Add all the attributes without attention to dupes initially, but track a flag to say whether there's been at least one call to `AddMultipleAttributes`
 * When closing an element/component, if the "is splatting" flag is set, then iterate backwards over the parameter frames to build a `Set<string>` containing all the parameter names. For each duplicate, don't delete it from the array - just overwrite it with a marker frame to say this frame should be ignored. For instance, it could be an attribute frame whose `AttributeName` is null, or it could be a new frame type entirely. Then we update any code that iterates over attribute frames so it knows to skip such items (there will only be a few places where we do this, as most just uses the custom iterator in `ParameterCollectionExtensions`). This reduces the O(N^2) to O(N).

Reply:
Hopefully then we wouldn't also need the massive number of unit tests dealing with deduplication of different types, as the deduplication logic would be the same code paths for all types.

Reply:
> Hopefully then we wouldn't also need the massive number of unit tests dealing with deduplication of different types, as the deduplication logic would be the same code paths for all types.

We could theoretically reduce the number of tests, but each code path where we call `AddAttribute` but don't create a frame needs to have special behaviour. This was my rationale for writing all of these tests, and I'm not totally convinced that it's safe to get rid of them.

I like your proposal and I'm going to try it out, but we're going to end up with *some* logic in each place where we have a *ghost frame*. This is the same concern that led me to writing a million tests. It would be really easy to introduce bugs in these cases - and its a concern that you need to be aware of every time we add another overload to AddAttribute.

Reply:
> Update: We could also make the "false bool" handling work by not only skipping adding the frame, but also by putting the name into the Set<string> at that point so that cleared values get overwritten during the deduplication pass if there is one.

This can work if we use a dictionary and track the position as well.

Imagine the following:
```
<div @foo disabled="@isDisabled" @bar>
```

When we process `disabled` we have *seen* an `AddMultipleAttributes` call (from `@foo`) so we track that we have seen `disabled`. 

Now when we process the `AddMultipleAttributes` for `@bar` we have already *seen* `disabled`. So if `@bar` also contains `disabled` then it would be skipped. Therefore set membership isn't quite enough, we need the position as well.

Reply:
A `Dictionary<string, int>` sounds good to me.

It also occurs to me that, once we know which frames are going to be ghosted (e.g., once we've actually ghosted them), we could also do a compaction step where in a single forwards linear pass, we remove all the ghosted items (like [this](https://www.geeksforgeeks.org/move-zeroes-end-array/)).

Then it's not necessary to find and update any other code that walks through attributes, so this *might* be less work overall (and avoids introducing a new concept we have to remember in the future).

I'm not fixated on whether or not we do it this way - up to you.